Welcome to Thomas Scott's COMP 483 Mini Project!

This project consists of a Python wrapper that automates several software tools used in bioinformatics.
It was made in and is intended to run in a UNIX terminal.
The tools used include kallisto, sleuth, Bowtie2, SPAdes, and BLAST+.
You will need to have these tools installed for the pipeline to run correctly.

A full list of tools to be installed and links to instructions are included here:

kallisto: https://pachterlab.github.io/kallisto/source

R: https://www.r-project.org/

sleuth (R package): https://pachterlab.github.io/sleuth/download

dplyr (R package): https://www.r-project.org/nosvn/pandoc/dplyr.html

Bowtie2: http://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#obtaining-bowtie-2

SPAdes: https://cab.spbu.ru/files/release3.15.1/manual.html

BLAST+: https://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE_TYPE=BlastDocs&DOC_TYPE=Download

Biopython: https://biopython.org/wiki/Download


This pipeline uses transcriptome data of HCMV (Human cytomegalovirus/Human herpesvirus 5) 2- and 6-days post-infection.
The pipeline produces a variety of outputs and writes important information to a log file titled 'miniProject.log'.
A 'miniProject.log' file is included in this repository as an example of the output generated by a full dataset pipeline run.

Aside from the Python wrapper ('MiniProject.py'), some other needed files are included.
miniProject_sleuth.R is the sleuth script ran during the pipeline, with miniProject_table.txt being used in that script.
Furthermore, a betaherpes.fasta file is included, so that a local BLAST+ database can be created for betaherpesvirinae towards the end of the pipeline.

As the pipeline requires a large amount of time to run, test data that is a small subset of reads from the input data is included.
If you would like to use the test data, simply include the testdata folder/directory in the same folder as the python wrapper ('MiniProject.py').
If you would like to run the full pipeline, you may move or delete the testdata folder.

Below are instructions on running the code once the required tools are installed:

1.) Be sure your present working directory is where you want to run the pipeline.

2.) Clone this Github repository to your local machine using the command "git clone https://github.com/tscott3LUC/483miniproject.git"

  a.) If you do not have Git/Github set up, check out this link to learn about how to get started: https://docs.github.com/en/github/getting-started-with-github/set-up-git
  
3.) You should now see a directory named '483miniProject_Thomas_Scott.' In this directory are all the files included in this repository. Change your working directory ('cd' command) to this directory.

4.) To run the pipeline, simply run the command 'python3 MiniProject.py'.

  a.) Again, this will run the test data, provided it has not been moved or deleted.
  
5.) Now you can just wait! The pipeline using test data should take less than 5 minutes. If you use the full dataset, it will take a few hours.

  a.) The script will create a new directory named 'miniProject_Thomas_Scott' - this is where all pipeline results will be generated.
  
  b.) Some files (miniProject_sleuth.R, miniProject_table.txt, betaherpes.fasta) will be copied into this directory for use during the pipeline, so you will see these in the directory as well.
  
6.) Important results will be written to 'miniProject.log' in the 'miniProject_Thomas_Scott' directory.

